src/dawn/Optimizer/PassSetBoundaryCondition.cpp

```diff
@@ -0,0 +1,337 @@
+                        visitor.reset();
+
+                        std::shared_ptr<Stmt>& stmt = statement->ASTStmt;
+
+                        stmt->accept(visitor);
+                        std::vector<std::shared_ptr<Stmt>> stencilCallWithBC_;
+                        stencilCallWithBC_.emplace_back(boundaryConditionCall);
+                        stencilCallWithBC_.emplace_back(test->second);
+
+                        for(auto& oldStencilCall : visitor.getStencilCallsToReplace()) {
+                          auto newBlockStmt = std::make_shared<BlockStmt>();
+                          std::copy(stencilCallWithBC_.begin(), stencilCallWithBC_.end(),
+                                    std::back_inserter(newBlockStmt->getStatements()));
+                          if(oldStencilCall == stmt) {
+                            // Replace the the statement directly
+                            DAWN_ASSERT(visitor.getStencilCallsToReplace().size() == 1);
+                            stmt = newBlockStmt;
+                          } else {
+                            // Recursively replace the statement
+                            replaceOldStmtWithNewStmtInStmt(stmt, oldStencilCall, newBlockStmt);
+                          }
+                        }
+                      }
+
+                      // The boundary condition is applied, the field is clean again
+                      dirtyFields.erase(originalID);
+                      // we add it to a vector for output
+                      boundaryConditionInserted_.push_back(originalID);
+                    } else {
+                      DAWN_ASSERT_MSG(false,
+                                      "Stencil Triggering the Boundary Condition is not called");
+                    }
+                  } else {
+                    DAWN_ASSERT_MSG(
+                        false,
+                        dawn::format("In stencil %s we need a halo update on field %s but no "
+                                     "boundary condition is set.\nUpdate the stencil (outside the "
+                                     "do-method) with a boundary condition that calls a "
+                                     "stencil_function, e.g \n'boundary_condition(zero(), %s);'\n",
+                                     stencilInstantiation->getName(),
+                                     stencilInstantiation->getOriginalNameFromAccessID(originalID),
+                                     stencilInstantiation->getOriginalNameFromAccessID(originalID))
+                            .c_str());
+                    dirtyFields.erase(originalID);
+                  }
+                }
+              }
+            }
+            // Any write-access requires a halo update once is is read off-center therefore we set
+            // the fields to modified
+            for(const auto& writeaccess : allWriteAccesses) {
+              int originalID = getOriginalID(writeaccess.first);
+              if(originalID != -1) {
```

Comment:
``` if(originalID == -1) continue```

Reply:
btw, in the future we should make `-1` a tag, enum or something rather than hardcoded